File: ts/test/abnf.test.js
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */
'use strict'

/*  abnf.test.js
 *  Round-trip test for the debug plugin's `abnf()` emitter.
 *
 *  HARD INDEPENDENCE CONSTRAINT: @tabnas/abnf must NOT be a dependency of
 *  @tabnas/debug. The emitter (src/debug.ts) reads ONLY the live engine
 *  and never imports abnf. abnf is used HERE, in the test only, and is
 *  loaded by SIBLING PATH (never via package.json) so it stays out of
 *  the dependency graph.
 *
 *  Round-trip criterion: for a sample ABNF A0,
 *    G1 = abnfConvert(A0) installed on a Tabnas instance;
 *    A1 = thatInstance.debug.abnf();
 *    G2 = abnfConvert(A1) on a fresh instance;
 *  G1 and G2 must RECOGNISE the sample inputs identically — same parse
 *  success/failure and same top `.rule` name. (ABNF has no actions, so
 *  parse-output values are out of scope.)
 */

const { describe, it } = require('node:test')
const assert = require('node:assert')
const path = require('node:path')

const { Tabnas } = require('@tabnas/parser')
const { Debug } = require('..')

// abnf, loaded by sibling PATH only — NOT a package dependency. The debug
// repo sits beside the abnf repo (`@tabnas/abnf`, in the `abnf` directory)
// in the tabnas multi-repo layout; resolve its built dist relative to this
// test file so the path stays correct regardless of where the repos are
// checked out.
const { abnfConvert } = require(
  require('path').resolve(__dirname, '..', '..', '..', 'abnf', 'ts', 'dist', 'abnf.js'),
)

// Recognise `input` with a compiled grammar `spec`. Returns a normalised
// result: { ok, rule } where `ok` reflects parse success and `rule` is
// the top node's grammar-rule tag (undefined when absent).
function recognise(spec, input) {
  try {
    const tn = new Tabnas().grammar(spec)
    const out = tn.parse(input)
    return { ok: true, rule: out && out.rule }
  } catch (e) {
    return { ok: false }
  }
}

// Assert the emitted text obeys the parts of RFC 5234 that were previously
// violated silently. @tabnas/abnf is lenient about both, so re-compiling is
// NOT evidence of validity — the trailing-`/` bug survived this whole suite
// because every check downstream went through the tolerant parser.
//
//   rulename    = ALPHA *(ALPHA / DIGIT / "-")
//   alternation = concatenation *(*c-wsp "/" *c-wsp concatenation)
//
// so `_gen1_star_x` is not a legal name, and `x = A x /` is not a legal body.
function assertRfc5234Shape(abnf1) {
  for (const line of abnf1.split('\n')) {
    if ('' === line.trim() || line.trim().startsWith(';')) continue

    assert.ok(
      !/\/\s*$/.test(line),
      'dangling `/` — every `/` needs a concatenation after it:\n' +
      line + '\n--- in ---\n' + abnf1,
    )

    const head = line.match(/^([^\s=]+)\s*=/)
    if (head) {
      assert.match(
        head[1],
        /^[A-Za-z][A-Za-z0-9-]*$/,
        'rulename is not ALPHA *(ALPHA / DIGIT / "-"):\n' +
        line + '\n--- in ---\n' + abnf1,
      )
    }
  }
}

// Assert the full round trip for one sample grammar over a set of inputs.
function assertRoundTrip(abnf0, inputs) {
  const g1 = abnfConvert(abnf0)

  const tn = new Tabnas()
  tn.use(Debug, { print: false, trace: false })
  tn.grammar(g1)
  const abnf1 = tn.debug.abnf()

  assert.strictEqual(typeof abnf1, 'string', 'abnf() returns a string')
  assert.ok(abnf1.length > 0, 'abnf() output is non-empty')
  assertRfc5234Shape(abnf1)

  let g2
  try {
    g2 = abnfConvert(abnf1)
  } catch (e) {
    assert.fail(
      'emitted ABNF did not re-compile:\n' + abnf1 + '\n' + e.message,
    )
  }

  for (const input of inputs) {
    const r1 = recognise(g1, input)
    const r2 = recognise(g2, input)
    assert.deepStrictEqual(
      r2,
      r1,
      'recognition mismatch for ' +
      JSON.stringify(input) +
      '\n  A0 = ' +
      JSON.stringify(abnf0) +
      '\n  A1 = ' +
      JSON.stringify(abnf1),
    )
  }
}

describe('abnf', () => {
  it('decorates an instance with abnf()', () => {
    const tn = new Tabnas()
    tn.use(Debug, { print: false, trace: false })
    assert.strictEqual(typeof tn.debug.abnf, 'function')
  })

  it('round-trips alternation', () => {
    assertRoundTrip('greet = "hi" / "hello"', ['hi', 'hello', 'nope', ''])
  })

  it('round-trips concatenation', () => {
    assertRoundTrip('pair = "a" "b"', ['ab', 'a', 'ba', ''])
  })

  it('round-trips a rule reference', () => {
    assertRoundTrip('top = greet\ngreet = "hi"', ['hi', 'no', ''])
  })

  it('round-trips a case-sensitive literal', () => {
    assertRoundTrip('g = %s"Hi"', ['Hi', 'hi', 'HI', ''])
  })

  it('round-trips a char-range', () => {
    assertRoundTrip('g = %x30-39', ['5', '0', 'a', ''])
  })

  // `char-val = DQUOTE *(%x20-21 / %x23-7E) DQUOTE`, so a token fixed to a
  // control character cannot be quoted — it has to come back as a num-val.
  // Quoting it produced `CRLF = "<CR>"`, an unterminated char-val.
  it('round-trips a control-character literal as %x, not a quoted char-val', () => {
    // `emit` is defined further down the describe body; `it` callbacks run
    // after that body completes, so it is initialised by the time this runs.
    const out = emit('csv = row *( CR row )\nrow = "x"\nCR = %x0D')
    assert.match(out, /^CR\s+= %x0D$/m, 'control char emitted as num-val:\n' + out)
    assertRfc5234Shape(out)
    assertRoundTrip(
      'csv = row *( CR row )\nrow = "x"\nCR = %x0D',
      ['x', 'x\rx', 'x\rx\rx', '', 'y'],
    )
  })

  // Extra coverage beyond the required minimum: these all round-trip.
  it('round-trips ref-only alternation (FIRST-set peek)', () => {
    assertRoundTrip(
      'top = a / b\na = "x"\nb = "y"',
      ['x', 'y', 'z', ''],
    )
  })

  it('round-trips repetition (star and plus)', () => {
    assertRoundTrip('rep = *"a"', ['', 'a', 'aa', 'b'])
    assertRoundTrip('rep = 1*"a"', ['', 'a', 'aa', 'b'])
  })

  // Repetition inside a group is what produces the deepest synthetic names
  // (`_gen2_star__gen1_group$alt0$step1`) — the shape most likely to emit
  // an illegal rulename.
  it('round-trips repetition inside a group', () => {
    assertRoundTrip(
      'list = "[" *( "," item ) "]"\nitem = "x"',
      ['[]', '[,x]', '[,x,x]', '[x]', '['],
    )
  })

  it('round-trips optional (group and prefix)', () => {
    assertRoundTrip('opt = ["a"]', ['', 'a', 'aa'])
    assertRoundTrip('m = ["x"] "y"', ['y', 'xy', 'x', 'xx'])
  })

  it('round-trips a grouped alternation', () => {
    assertRoundTrip('g = ("a" / "b") "c"', ['ac', 'bc', 'c'])
  })

  it('round-trips a multi-rule grammar with mixed terminals', () => {
    assertRoundTrip(
      'uri = scheme ":" path\n' +
      'scheme = "http" / "https"\n' +
      'path = "/a" / "/b"',
      ['http:/a', 'https:/b', 'ftp:/a', ':'],
    )
  })

  // Explicit structural checks on the emitted ABNF (the round-trip tests
  // above verify recognition equivalence; these pin the exact folded shape,
  // matching the Go TestAbnfFoldsSyntheticOptional / TestAbnfKeepsRepetition
  // tests). emit() compiles A0 via abnf and returns what debug re-emits.
  const emit = (abnf0) => {
    const tn = new Tabnas()
    tn.use(Debug, { print: false, trace: false })
    tn.grammar(abnfConvert(abnf0))
    return tn.debug.abnf()
  }

  it('folds a synthetic optional back to [ … ] (no _gen leaks)', () => {
    const out = emit('add = NR [ PL add ]\nPL = "+"')
    assert.ok(
      out.split('\n').includes('add = NR [ PL add ]'),
      'optional folded to `NR [ PL add ]`:\n' + out,
    )
    // Matches the sanitised spelling too: `_gen1_star_x` now emits as
    // `r-gen1-star-x`, so a bare /_gen/ would pass without testing anything.
    assert.ok(
      !/[_-]gen\d/.test(out),
      'no synthetic gen production leaked:\n' + out,
    )
  })

  it('keeps repetition as a production (does not fold *)', () => {
    const out = emit('rep = *PL\nPL = "+"')
    assert.ok(
      /^r-gen\d+-star-PL = /m.test(out),
      'star kept as its own production:\n' + out,
    )
    // Zero-or-more IS the empty alternative, and it is rendered as `[ … ]`.
    // NOT as a trailing `/`: RFC 5234's `alternation` requires a
    // concatenation after every `/`, so `x = PL x /` is a syntax error that
    // @tabnas/abnf happens to accept and other ABNF tools do not.
    assert.ok(
      /^r-gen\d+-star-PL = \[ PL r-gen\d+-star-PL \]$/m.test(out),
      'zero-or-more empty alternative rendered as `[ … ]`:\n' + out,
    )
    assert.ok(!/\/\s*$/m.test(out), 'no dangling `/`:\n' + out)
  })

  it('describe() includes an ABNF section', () => {
    const tn = new Tabnas()
    tn.use(Debug, { print: false, trace: false })
    tn.grammar(abnfConvert('greet = "hi" / "hello"'))
    const desc = tn.debug.describe()
    assert.ok(desc.includes('========= ABNF ========='), 'has ABNF header')
    assert.ok(desc.includes('greet = HI / HELLO'), 'has emitted ABNF rule')
    assert.ok(/\bHI\b\s*=\s*"hi"/.test(desc), 'has token definition')
  })
})

File: rs/tests/abnf_test.rs
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

//! Whole-emitter tests for `abnf`, mirroring the `TestAbnf*` set in
//! `../../go/debug_test.go` and the cases in `../../ts/test/abnf.test.js`.
//!
//! What the shared `../../test/spec/abnf.tsv` fixture already pins for
//! every runtime is NOT repeated here: alternation, `%s` literals, the
//! synthetic-optional fold, the rule/token name collision and the
//! prose-val fallback all have rows there. What is left are the grammar
//! SHAPES the fixture registry does not hold, each of which needs a
//! purpose-built instance.
//!
//! The canonical TypeScript proves most of these by round-tripping
//! through `@tabnas/abnf`. That suite cannot run here, or in Go: the
//! emitter must never gain an ABNF dependency, and neither port may take
//! one even in a test without putting that independence claim in doubt.
//! So both ports pin the same shapes by asserting the emitted text, and
//! this file is the Rust half.

mod common;

use tabnas::{AltSpec, Tabnas, Tin};
use tabnas_debug::abnf;

/// Resolve a built-in token identity, failing loudly if the engine has
/// renamed it.
fn token(parser: &Tabnas, name: &str) -> Tin {
    parser
        .options
        .token(name)
        .unwrap_or_else(|| panic!("the engine has no {name} token"))
}

/// The parts of RFC 5234 the emitter has violated before:
///
/// ```text
/// rulename    = ALPHA *(ALPHA / DIGIT / "-")
/// alternation = concatenation *(*c-wsp "/" *c-wsp concatenation)
/// ```
///
/// so `_gen1_star_x` is not a legal name, and `x = A x /` is not a legal
/// body. Mirrors `assertRfc5234Shape` in the TypeScript and Go suites.
fn assert_rfc5234_shape(out: &str) {
    for line in out.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with(';') {
            continue;
        }
        assert!(
            !trimmed.ends_with('/'),
            "dangling `/`, every `/` needs a concatenation after it:\n{line}\n--- in ---\n{out}"
        );
        if let Some(head) = head_name(line) {
            assert!(
                is_legal_rulename(head),
                "rulename is not ALPHA *(ALPHA / DIGIT / \"-\"):\n{line}\n--- in ---\n{out}"
            );
        }
    }
}

/// The rule name a production line defines, i.e. everything before the
/// first `=`, when the line has one. Mirrors the `^([^\s=]+)\s*=` the
/// other two suites use.
fn head_name(line: &str) -> Option<&str> {
    let (head, _) = line.split_once('=')?;
    let head = head.trim();
    if head.is_empty() || head.contains(char::is_whitespace) {
        return None;
    }
    Some(head)
}

fn is_legal_rulename(name: &str) -> bool {
    let mut chars = name.chars();
    chars
        .next()
        .is_some_and(|first| first.is_ascii_alphabetic())
        && chars.all(|ch| ch.is_ascii_alphanumeric() || '-' == ch)
}

/// The shape the ABNF compiler emits for a repetition (`rep = *T`): a
/// `_gen…_star_…` production whose empty open alternative marks it
/// zero-or-more. Unlike `[ … ]`, repetition uses a probe-optimised
/// subgraph that does not reconstruct as `*( … )` reliably, so the
/// emitter must KEEP it as a production rather than fold it.
fn star_grammar() -> Tabnas {
    let mut parser = Tabnas::new();
    parser.options.rule.start = "rep".into();
    let plus = parser.token_with_source("#T", "+");

    parser.define_rule("rep", |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            p: Some("_gen1_star_T".into()),
            ..Default::default()
        });
        spec.add_close(AltSpec::new());
    });
    parser.define_rule("_gen1_star_T", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![plus]],
            ..Default::default()
        });
        // The empty alternative is what makes it zero-or-more.
        spec.add_open(AltSpec::new());
        spec.add_close(AltSpec::new());
    });
    parser
}

/// A `_gen…_star_…` synthetic is NOT folded. It stays a bareword
/// reference in its parent and is emitted as its own production. Beyond
/// the folding rule this pins two things RFC 5234 requires: the empty
/// alternative renders as `[ … ]` rather than a trailing `/`, and the
/// synthetic name is sanitised to a legal rulename, so `_gen1_star_T`
/// comes out as `r-gen1-star-T`.
#[test]
fn abnf_keeps_a_repetition_production() {
    let out = abnf(&star_grammar());
    assert_eq!(
        out, "rep = r-gen1-star-T\nr-gen1-star-T = [ T ]\n\nT = \"+\"",
        "repetition mismatch:\n{out}"
    );
    assert!(
        !out.contains("_gen"),
        "a synthetic _gen name leaked:\n{out}"
    );
    assert_rfc5234_shape(&out);
}

/// The num-val fallback for a fixed token whose literal cannot go inside
/// quotes. RFC 5234 has `char-val = DQUOTE *(%x20-21 / %x23-7E) DQUOTE`,
/// so a control character, a DQUOTE, or anything above `%x7E` has to come
/// back as `%xNN`. Quoting a CR emitted `CR = "<CR>"`, an unterminated
/// char-val.
#[test]
fn abnf_renders_an_unquotable_literal_as_a_num_val() {
    let mut parser = Tabnas::new();
    parser.options.rule.start = "top".into();
    let cr = parser.token_with_source("#CR", "\r");
    let dq = parser.token_with_source("#DQ", "\"");

    parser.define_rule("top", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![cr], vec![dq]],
            ..Default::default()
        });
        spec.add_close(AltSpec::new());
    });

    let out = abnf(&parser);
    for want in ["CR = %x0D", "DQ = %x22"] {
        assert!(out.contains(want), "missing {want:?} in:\n{out}");
    }
    assert_rfc5234_shape(&out);
}

/// The multi-character, padding and non-ASCII cases of the num-val
/// rendering, driven through the public surface rather than the private
/// helper.
#[test]
fn abnf_num_val_covers_multi_char_padding_and_non_ascii() {
    let mut parser = Tabnas::new();
    parser.options.rule.start = "top".into();
    let crlf = parser.token_with_source("#CRLF", "\r\n");
    let nul = parser.token_with_source("#NUL", "\0");
    let acc = parser.token_with_source("#ACC", "\u{e9}");
    let tab = parser.token_with_source("#TAB", "\t");

    parser.define_rule("top", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![crlf], vec![nul], vec![acc], vec![tab]],
            ..Default::default()
        });
        spec.add_close(AltSpec::new());
    });

    let out = abnf(&parser);
    for want in [
        "CRLF = %x0D.0A", // multi-character, dot-concatenated
        "NUL  = %x00",    // zero-padded to two hex digits
        "ACC  = %xE9",    // non-ASCII
        "TAB  = %x09",    // control character
    ] {
        assert!(out.contains(want), "missing {want:?} in:\n{out}");
    }
    assert_rfc5234_shape(&out);
}

/// A rule whose only open alternative is empty, but which has a close
/// continuation, emits the continuation alone.
/// `option = "[" *c-wsp alternation *c-wsp "]"` and `alternation` needs
/// at least one concatenation, so `[ ]` is not a legal option. This used
/// to emit `inner = [  ] AA`.
#[test]
fn abnf_emits_a_continuation_alone_not_an_empty_option() {
    let mut parser = Tabnas::new();
    parser.options.rule.start = "top".into();
    let xa = parser.token_with_source("#XA", "a");
    let zz = token(&parser, "#ZZ");

    parser.define_rule("top", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            p: Some("inner".into()),
            ..Default::default()
        });
        spec.add_close(AltSpec {
            s: vec![vec![zz]],
            ..Default::default()
        });
    });
    parser.define_rule("inner", move |spec| {
        spec.clear();
        spec.add_open(AltSpec::new());
        spec.add_close(AltSpec {
            s: vec![vec![xa]],
            ..Default::default()
        });
        spec.add_close(AltSpec {
            s: vec![vec![zz]],
            ..Default::default()
        });
    });

    let out = abnf(&parser);
    assert!(!has_empty_option(&out), "an empty `[ ]` option in:\n{out}");
    assert_rfc5234_shape(&out);
}

/// `[` then only whitespace then `]`, the shape the other two suites
/// match with the regex `\[\s*\]`.
fn has_empty_option(out: &str) -> bool {
    let bytes: Vec<char> = out.chars().collect();
    for (index, ch) in bytes.iter().enumerate() {
        if '[' != *ch {
            continue;
        }
        let mut cursor = index + 1;
        while cursor < bytes.len() && bytes[cursor].is_whitespace() {
            cursor += 1;
        }
        if cursor < bytes.len() && ']' == bytes[cursor] {
            return true;
        }
    }
    false
}

/// RFC 5234 §2.1: rule names are case-insensitive, so `Foo-Bar` and
/// `foo-bar` are ONE rule. Sanitising `foo_bar` beside a reserved
/// `Foo-Bar` used to emit two definitions of the same rule.
#[test]
fn abnf_rulename_collisions_are_case_insensitive() {
    let mut parser = Tabnas::new();
    parser.options.rule.start = "top".into();
    let xa = parser.token_with_source("#XA", "a");
    let xb = parser.token_with_source("#XB", "b");
    let zz = token(&parser, "#ZZ");

    parser.define_rule("top", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![xa]],
            p: Some("Foo-Bar".into()),
            ..Default::default()
        });
        spec.add_open(AltSpec {
            s: vec![vec![xb]],
            p: Some("foo_bar".into()),
            ..Default::default()
        });
        spec.add_close(AltSpec {
            s: vec![vec![zz]],
            ..Default::default()
        });
    });
    parser.define_rule("Foo-Bar", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![xa]],
            ..Default::default()
        });
        spec.add_close(AltSpec::new());
    });
    parser.define_rule("foo_bar", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![xb]],
            ..Default::default()
        });
        spec.add_close(AltSpec::new());
    });

    let out = abnf(&parser);
    let mut seen: Vec<String> = Vec::new();
    for line in out.lines() {
        let Some(head) = head_name(line) else {
            continue;
        };
        let key = head.to_lowercase();
        assert!(
            !seen.contains(&key),
            "two productions define the same rule (case-insensitively): {head:?} in:\n{out}"
        );
        seen.push(key);
    }
    assert_rfc5234_shape(&out);
}

/// The skip branch of an optional arrives as a FIRST-set-guarded epsilon:
/// an alt carrying the FOLLOW token in `s` with `b` set and NO
/// push/replace target. The token is lookahead, matched to choose the
/// branch and then backtracked, so the alt consumes nothing
/// (`s.len() - b == 0`).
///
/// The emitter used to skip `s` only when `b` was set AND the alt pushed
/// or replaced, so this shape rendered as a CONSUMING alternative:
/// `top = [ X "@" ] Y` came back as `top = [ X T / Y ] Y`, where the
/// optional could swallow the follow and leave nothing for the trailing
/// `Y`.
#[test]
fn abnf_does_not_render_a_follow_guarded_epsilon_as_an_alternative() {
    let mut parser = Tabnas::new();
    parser.options.rule.start = "top".into();
    let at = parser.token_with_source("#T", "@");
    let yy = parser.token_with_source("#Y", "b");
    let xx = parser.options.register_token("#X");

    parser.define_rule("top", |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            p: Some("_gen2_opt__gen1_group".into()),
            ..Default::default()
        });
        spec.add_close(AltSpec {
            r: Some("top$step1".into()),
            ..Default::default()
        });
    });
    parser.define_rule("top$step1", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![yy]],
            ..Default::default()
        });
    });
    parser.define_rule("_gen1_group", move |spec| {
        spec.clear();
        spec.add_open(AltSpec {
            s: vec![vec![xx], vec![at]],
            ..Default::default()
        });
    });
    parser.define_rule("_gen2_opt__gen1_group", move |spec| {
        spec.clear();
        // Take the optional: peek #X, the pushed group consumes it.
        spec.add_open(AltSpec {
            s: vec![vec![xx]],
            b: 1,
            p: Some("_gen1_group".into()),
            ..Default::default()
        });
        // Skip it: peek the FOLLOW #Y and consume nothing. The shape at
        // issue.
        spec.add_open(AltSpec {
            s: vec![vec![yy]],
            b: 1,
            ..Default::default()
        });
        // Bare epsilon.
        spec.add_open(AltSpec::new());
        spec.add_close(AltSpec::new());
    });

    let out = abnf(&parser);
    for line in out.lines() {
        if !line.trim_start().starts_with("top") {
            continue;
        }
        // The follow token must not appear as an alternative INSIDE the
        // option. `top = [ X T ] Y` is right; `top = [ X T / Y ] Y` is
        // the defect.
        if let Some(close) = line.find(']') {
            assert!(
                !line[..close].contains('/'),
                "a follow-guarded epsilon rendered as a consuming alternative:\n{line}\n--- in ---\n{out}"
            );
        }
    }
    assert_rfc5234_shape(&out);
}

/// Every grammar in the shared registry emits ABNF that obeys the two
/// RFC 5234 shapes above. The `.tsv` fixture pins the exact bytes for
/// each; this pins the property, so a new fixture grammar cannot land an
/// illegal name or a dangling `/` merely by having its row regenerated
/// from the canonical runtime.
#[test]
fn every_fixture_grammar_emits_rfc5234_shaped_abnf() {
    for name in ["bare", "add", "greet", "collide"] {
        let parser = common::fixture::build(name).expect("a known grammar");
        assert_rfc5234_shape(&abnf(&parser));
    }
}
